### PR TITLE
RFC: Avoid setting the RPATH for LLVM

### PR DESCRIPTION
**Pull request by ararslan**
Some of the changes in #21788 adversely affected Linux. This partially reverts/restructures some of the changes.

### CODE REVIEW

**Comment by ararslan** [comment]
@jlbuild !nuke

**Comment by jlbuild** [comment]
Got it, building https://github.com/JuliaLang/julia/commit/1c973df4455d0eb558c687d4570815c7a0a04ace

I will edit this comment as the build jobs complete.

**Review by vtjnash — APPROVED** [review]
lgtm

**Comment by vtjnash** [comment]
I don't know if this was related to any of your PRs, but I noticed we're creating some of these variables as normal recursively expanded ones, instead of as simply-expanded ones. We may want to add this patch to declare these variables as being of type `:=` before using them.
```
--- a/deps/llvm.mk
+++ b/deps/llvm.mk
@@ -60,6 +60,12 @@ else
 LLVM_TARGETS := host;NVPTX
 endif
 
+LLVM_CFLAGS :=
+LLVM_CXXFLAGS :=
+LLVM_CPPFLAGS :=
+LLVM_LDFLAGS :=
+LLVM_CMAKE :=
+
 # Allow adding LLVM specific flags
 LLVM_CFLAGS += $(CFLAGS)
 LLVM_CXXFLAGS += $(CXXFLAGS)
```

**Comment by ararslan** [comment]
No, I think the variable declaration was a preexisting condition. I've added your patch as a separate commit. Thanks!

**Comment by tkelman** [comment]
Why is llvm picky about this? leave a comment by the CMAKE_COMMON saying not to put anything rpath related there for this reason, so we aren't tempted to do this again by accident?

**Comment by ararslan** [comment]
> Why is llvm picky about this?

¯\\\_(ツ)\_/¯ It isn't on FreeBSD, which is why I never noticed this

> leave a comment [...] so we aren't tempted to do this again by accident?

Good idea, will do once CI is done so I don't end up just queuing up a ton of builds here

**Review by Keno — APPROVED** [review]

File: deps/tools/common.mk [hunk 1]
```diff
@@ -40,16 +40,14 @@ CMAKE_COMMON += -DCMAKE_RC_COMPILER="$$(which $(CROSS_COMPILE)windres)"
...
-ifneq (,$(findstring $(OS),Linux FreeBSD))
-INSTALL_RPATH := "\$$ORIGIN"
-# Part of the FreeBSD libgcc_s kludge
+# NOTE: Do not make RPATH changes in CMAKE_COMMON on platforms other than FreeBSD, since
+# it will make its way into the LLVM build flags, and LLVM is picky about RPATH (though
+# apparently not on FreeBSD). Ref PR #22352
 ifeq ($(OS),FreeBSD)
 ifneq ($(GCCPATH),)
-INSTALL_RPATH := "\$$ORIGIN:$(GCCPATH)"
+CMAKE_COMMON += -DCMAKE_INSTALL_RPATH="\$$ORIGIN:$(GCCPATH)"
```
Comment (ararslan) [inline deps/tools/common.mk]: This actually ends up duplicating the RPATH specification on FreeBSD, though I nuked everything, rebuilt from scratch, and tested everything, and things are fine.

Comment (ararslan) [inline deps/tools/common.mk]: image

**Pull request merged**